### PR TITLE
[Finishes PT-153539819] Refactor aec_tx_sign to being a real datatype

### PR DESCRIPTION
**Pull request by ThomasArts**
aec_tx_sign lends itself to being a pure datatype, which makes maintenance of our code base easier and helps testing by mocking only one part and no other modules.

### CODE REVIEW

File: apps/aecore/src/aec_tx_sign.erl [hunk 1]
```diff
@@ -1,23 +1,100 @@
...
+%% determine its own signers by the transaction callback 'signers'. Since we do not
+%% want to depend upon transaction types in this module, the user of
+%% {@module} should first obtain the signers of the transaction and then call this
+%% {@link sign/2} with these signers. There is a {@link sign/3} function that can sign
+%% with respect to a certain block height. This is handy whenever the governance
+%% variables on what crypto to use would change.
+
 %% API
--export([data/1,
+-export([sign/2,
```
Comment (gorillainduction) [inline apps/aecore/src/aec_tx_sign.erl]: While you're at it, can you sort the export entries, and either make them one entry or explain what the difference is.

File: apps/aecore/src/aec_tx_sign.erl [hunk 2]
```diff
@@ -1,23 +1,100 @@
...
+         serialize/1,
          serialize_to_binary/1,
          deserialize/1,
          deserialize_from_binary/1]).
 
 -include("common.hrl").
--include("txs.hrl").
 
+-record(signed_tx, {
+          data                       :: term(),
```
Comment (gorillainduction) [inline apps/aecore/src/aec_tx_sign.erl]: I think we can find out what this is. Probably a couple of transaction types.

File: apps/aecore/src/aec_tx_sign.erl [hunk 3]
```diff
@@ -1,23 +1,100 @@
...
+%% parameters.
+sign(Tx, PrivKeys) ->
+  sign(Tx, PrivKeys, #{}).
+
+%% @doc Given a transaction Tx, a private key and a crypto map, 
+%% return the cryptographically signed transaction.
+%% @equiv sign(Tx, [PrivKey], CryptoMap).
+sign(Tx, PrivKey, CryptoMap) when is_binary(PrivKey) ->
+  sign(Tx, [PrivKey], CryptoMap);
+sign(Tx, PrivKeys, CryptoMap) when is_list(PrivKeys) ->
```
Comment (gorillainduction) [inline apps/aecore/src/aec_tx_sign.erl]: Wouldn't it make sense to make sure that the CryptoMap contains the obligatory fields?

In sign/2 we could seed the argument with the default map.

File: apps/aecore/src/aec_tx_sign.erl [hunk 4]
```diff
@@ -1,23 +1,100 @@
...
+%% @doc Given a transaction Tx, a private key or list of keys, 
+%% return the cryptographically signed transaction using the default crypto
+%% parameters.
+sign(Tx, PrivKeys) ->
+  sign(Tx, PrivKeys, #{}).
+
+%% @doc Given a transaction Tx, a private key and a crypto map, 
+%% return the cryptographically signed transaction.
+%% @equiv sign(Tx, [PrivKey], CryptoMap).
+sign(Tx, PrivKey, CryptoMap) when is_binary(PrivKey) ->
```
Comment (hanssv) [inline apps/aecore/src/aec_tx_sign.erl]: Missing type spec?

File: apps/aecore/test/aec_mining_tests.erl [hunk 5]
```diff
@@ -31,10 +31,12 @@ mine_block_test_() ->
          fun() ->
                  TopBlock = #block{height = 0,
                                    target = ?HIGHEST_TARGET_SCI},
-                 meck:expect(aec_pow, pick_nonce, 0, 9),
```
Comment (gorillainduction) [inline apps/aecore/test/aec_mining_tests.erl]: I don't see how this is related to this PR.

If the behavior for this test case changed due to this PR, I think we should find the reason for this.

If this is a general test case improvement, make another PR.

File: apps/aecore/test/aec_blocks_tests.erl [hunk 6]
```diff
@@ -71,23 +71,27 @@ network_serialization_test_() ->
...
+             MalformedTxs = [SignedCoinbase, aec_tx_sign:sign(#coinbase_tx{account = <<"malformed_account">>}, <<"key">>)],
              {ok, MalformedTree} = aec_txs_trees:new(MalformedTxs),
              {ok, MalformedRootHash} = aec_txs_trees:root_hash(MalformedTree),
              Block = #block{txs = CorrectTxs, txs_hash = MalformedRootHash},
              ?assertEqual({error, malformed_txs_hash}, ?TEST_MODULE:validate(Block))
      end,
      fun() ->
-             SignedCoinbase = #signed_tx{data = #coinbase_tx{}},
+             SignedCoinbase = aec_tx_sign:sign(#coinbase_tx{}, <<"key">>),
+             %% This is weird, we test that a coinbase that is signed with a key that is not the account key does validate!!
```
Comment (hanssv) [inline apps/aecore/test/aec_blocks_tests.erl]: Aren't we testing that nothing bad happens if the coinbase tx is bad? The tree will not contain the coinbase tx in this case I think!?

File: apps/aecore/test/aec_mining_tests.erl [hunk 7]
```diff
@@ -161,8 +178,9 @@ setup(PoWMod) ->
     meck:expect(aec_tx, apply_signed, 3, {ok, Trees}),
     meck:expect(aec_keys, pubkey, 0, {ok, ?TEST_PUB}),
     meck:expect(aec_keys, sign, 1,
-                {ok, #signed_tx{data = #coinbase_tx{account = <<"pubkey">>},
-                                signatures = [<<"sig1">>]}}).
+                {ok, aec_tx_sign:sign(#coinbase_tx{account = <<"pubkey">>}, <<"sig1">>)}).
+%% #signed_tx{data = #coinbase_tx{account = <<"pubkey">>},
```
Comment (gorillainduction) [inline apps/aecore/test/aec_mining_tests.erl]: Remove commented out code.

File: apps/aecore/test/aec_mining_tests.erl [hunk 8]
```diff
@@ -31,10 +31,12 @@ mine_block_test_() ->
          fun() ->
                  TopBlock = #block{height = 0,
                                    target = ?HIGHEST_TARGET_SCI},
-                 meck:expect(aec_pow, pick_nonce, 0, 9),
-
-                 {ok, BlockCandidate, Nonce} = ?TEST_MODULE:create_block_candidate(TopBlock, []),
-                 {ok, Block} = ?TEST_MODULE:mine(BlockCandidate, Nonce),
+                 %% Try enough nonces to find a solution.
```
Comment (hanssv) [inline apps/aecore/test/aec_mining_tests.erl]: It should be deterministic, or we don't want it as a test... We don't want to loop and try many nonces here.
We are using aec_test_utils:mock_fast_and_deterministic_cuckoo_pow() in the cuckoo case.

File: apps/aecore/test/aec_mining_tests.erl [hunk 9]
```diff
@@ -79,8 +94,9 @@ difficulty_recalculation_test_() ->
                  meck:expect(aec_blocks, new, 3,
                              #block{height = 30,
                                     target = ?HIGHEST_TARGET_SCI,
-                                    txs = [#signed_tx{data = #coinbase_tx{account = <<"pubkey">>},
-                                                      signatures = [<<"sig1">>]}],
+                                    txs = [aec_tx_sign:sign(#coinbase_tx{account = <<"pubkey">>}, <<"sig1">>)],
+                                    %% [#signed_tx{data = #coinbase_tx{account = <<"pubkey">>},
```
Comment (hanssv) [inline apps/aecore/test/aec_mining_tests.erl]: Commented out code..

File: apps/aecore/test/aec_tx_pool_tests.erl [hunk 10]
```diff
@@ -175,24 +177,25 @@ a_signed_tx(Sender, Recipient, Nonce, Fee) ->
...
     try
         [{_, PrivKey}] = ets:lookup(?TAB, PubKey),
         Signers = aec_tx:signers(Tx),
         true = lists:member(PubKey, Signers),
-        Bin = aec_tx:serialize_to_binary(Tx),
-        Sig = crypto:sign(algo(), digest(), Bin, [PrivKey, curve()]),
-        {ok, #signed_tx{data = Tx,
-                        signatures = [Sig]}}
+        {ok, aec_tx_sign:sign(Tx, PrivKey)}
+        %% Bin = aec_tx:serialize_to_binary(Tx),
```
Comment (hanssv) [inline apps/aecore/test/aec_tx_pool_tests.erl]: Commented out code

File: apps/aecore/include/txs.hrl [hunk 11]
```diff
@@ -1,8 +1,4 @@
--record(signed_tx, {
-          data                       :: term(),
-          signatures = ordsets:new() :: ordsets:ordset(binary())}).
--type(signed_tx() :: #signed_tx{}).
-
+ %% opaque type for data structure signed_tx
```
Comment (hanssv) [inline apps/aecore/include/txs.hrl]: This comment is odd, types should not be in .hrl files so moving it is right but leaving this comment behind doesn't really help anyone, or?!

File: apps/aecore/test/aec_tx_pool_tests.erl (hunk 10 above)
Comment (gorillainduction) [inline apps/aecore/test/aec_tx_pool_tests.erl]: Remove commented out code.

File: apps/aecore/test/aec_tx_pool_tests.erl [hunk 12]
```diff
@@ -175,24 +177,25 @@ a_signed_tx(Sender, Recipient, Nonce, Fee) ->
...
+        %%                 signatures = [Sig]}}
     catch
         error:Err ->
             erlang:error({Err, erlang:stacktrace()})
     end.
 
-algo()   -> ecdsa.
-digest() -> sha256.
-curve()  -> crypto:ec_curve(secp256k1).
+%% algo()   -> ecdsa.
```
Comment (gorillainduction) [inline apps/aecore/test/aec_tx_pool_tests.erl]: Remove commented out code.

File: apps/aecore/src/aec_tx_sign.erl (hunk 1 above)
Comment (ThomasArts) [reply apps/aecore/src/aec_tx_sign.erl]: I dislike the serialize and unserialize to be exported, but need to investigate who is using it.

File: apps/aecore/test/aec_mining_tests.erl (hunk 5 above)
Comment (ThomasArts) [reply apps/aecore/test/aec_mining_tests.erl]: The reason was that crypto adds salt and therefore has a non-deterministic result on computing the signed tx. I did not want to hard-code the signed_tx, but realize that I have to.

File: apps/aecore/src/aec_tx_sign.erl (hunk 2 above)
Comment (ThomasArts) [reply apps/aecore/src/aec_tx_sign.erl]: Actually, signed_tx should be a parameterized type signed_tx(tx_coinbase()) for example. But a bit too much to fix that in this PT

File: apps/aecore/src/aec_tx_sign.erl (hunk 3 above)
Comment (ThomasArts) [reply apps/aecore/src/aec_tx_sign.erl]: The crypto map should come from governance settings. Plan to do that in a separate refactoring
